background.ts

```diff
@@ -749,17 +764,121 @@ function newK8sManager() {
 class BackgroundCommandWorker implements CommandWorkerInterface {
   getSettings() {
     return JSON.stringify(cfg, undefined, 2);
   }
 
-  requestShutdown() {
+  /**
+   * On failure, returns status 400, with a human-readable error message.
+   * @param newSettings
+   */
+  async updateSettings(newSettings: UpdatableSettings): Promise<[string, string]> {
+    const newConfig: RecursivePartial<settings.Settings> = {};
+    let desiredVersion: string;
+    let versions: string[] = [];
+    let m: any;
+    let needToUpdate = false;
+
+    for (const k in newSettings) {
+      let v: string|boolean = newSettings[k];
+      let desiredEngine: settings.ContainerEngine;
+
+      switch (k) {
+      case 'kubernetes-version':
+        m = /^v?(.+?)(?:\+k3s\d+)?$/.exec(v as string);
+        if (!m) {
+          return ['', `not a valid Kubernetes version: <${ v }`];
+        }
+        desiredVersion = m[1];
+        versions = (await k8smanager.availableVersions).map(entry => entry.version.version);
+        if (versions.length === 0) {
+          return ['', 'no versions of Kubernetes were found'];
+        } else if (!versions.includes(desiredVersion)) {
+          return ['', `Kubernetes version ${ desiredVersion } not found`];
+        }
+        if (cfg.kubernetes.version !== desiredVersion) {
+          needToUpdate = true;
+          _.merge(newConfig, { kubernetes: { version: v as string } });
+        }
+        break;
+
+      case 'kubernetes-enabled':
+        if (typeof (v) !== 'boolean') {
+          switch (v.toLowerCase()) {
+          case 'true':
+          case 'yes':
+          case 'on':
+            v = true;
+            break;
+          case 'false':
+          case 'no':
+          case 'off':
+            v = false;
+            break;
+          default:
+            return ['', `invalid value for kubernetes-enabled: <${ v }>`];
+          }
+        }
+        if (cfg.kubernetes.enabled !== v) {
+          needToUpdate = true;
+          _.merge(newConfig, { kubernetes: { enabled: v } });
+        }
+        break;
+
+      case 'container-engine':
+        switch (v) {
+        case 'containerd':
+          desiredEngine = settings.ContainerEngine.CONTAINERD;
+          break;
+        case 'moby':
+          desiredEngine = settings.ContainerEngine.MOBY;
+          break;
+        case 'docker':
+          desiredEngine = settings.ContainerEngine.MOBY;
+          break;
+        default:
+          return ['', `Invalid value for container-engine: <${ v }>`];
+        }
+        if (cfg.kubernetes.containerEngine !== (v as settings.ContainerEngine)) {
+          needToUpdate = true;
+          _.merge(newConfig, { kubernetes: { containerEngine: desiredEngine } });
+        }
+        break;
+
+      default:
+        return ['', `Unrecognized setting: ${ k }`];
+      }
+    }
+    if (needToUpdate) {
+      writeSettings(newConfig);
+      // cfg is a global, and at this point newConfig has been merged into it :(
```

Comment:
Well, in the other places `writeSettings` is invoked because of a user's action in the UI. Here it needs to get the UI to update its view. So, not really.

Fortunately, updating the UI doesn't trigger another event that leads to calling `writeSettings` again, so we don't run the code an extra time.